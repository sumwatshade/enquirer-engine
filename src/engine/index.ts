export * from './OutputAdapter'
export * from './Engine'