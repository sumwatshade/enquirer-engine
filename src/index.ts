export * from "./validator/validator";
export * from './engine/Engine'
