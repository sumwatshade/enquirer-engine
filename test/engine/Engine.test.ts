import { ISurveySchema, IEngineOptions, EnquirerEngine, IEngineOutput } from '../../src'

import enquirer from 'enquirer';
import { mocked } from 'ts-jest/utils';
import * as uuid from 'uuid'
import mockdate from 'mockdate'

jest.mock('enquirer', () => ({
    prompt: jest.fn()
}));

jest.mock('uuid')

const mockEnquirer = mocked(enquirer, true);
const mockUuid = mocked(uuid, true);

describe("Engine", () => {
    let surveySchema: ISurveySchema, engineOptions: IEngineOptions

    beforeEach(() => {

        mockdate.set(1434320025275)

        engineOptions = {
            outputFormat: 'json-raw',
            userId: 'test-user'
        }

        surveySchema = {
            id: 'my-test-schema',
            createdAt: new Date('2020-04-04').toDateString(),
            prompts: [{
                type: 'scale',
                name: "My prompt",
                margin: [0, 1, 1, 2],
                message: "This is my prompt",
                scale: [
                    { "name": "1", "message": "Super gross!" },
                    { "name": "2", "message": "Kinda gross." },
                    { "name": "3", "message": "This would be alright." },
                    { "name": "4", "message": "Oh I want this." },
                    { "name": "5", "message": "I NEEEED IT!" }
                ],
                choices: [
                    {
                        "name": "inout",
                        "message": "In N Out"
                    },
                    {
                        "name": "tacobell",
                        "message": "Taco Bell"
                    },
                    {
                        "name": "pizza",
                        "message": "Howie's Artisan Pizza"
                    }
                ]
            }]
        }

        mockEnquirer.prompt.mockResolvedValue({
            tacobell: 2,
            pizza: 1,
            inout: 3
        })

    });

    afterEach(() => {
        mockdate.reset()
    })

    it("Will run a scale-based survey", async () => {
        const engine = new EnquirerEngine(engineOptions);

        const result: IEngineOutput = (await engine.run(surveySchema)) as IEngineOutput

        

        expect(result.id).toBe(surveySchema.id)
        expect(result.userId).toBe(engineOptions.userId)
        expect(result).toMatchSnapshot()

    });
    it("Will default a user ID if none are provided", async () => {
        delete engineOptions.userId;
        
        mockUuid.v4.mockReturnValue('hashValue')
        
        const engine = new EnquirerEngine(engineOptions);
        
        const result: IEngineOutput = await engine.run(surveySchema) as IEngineOutput
        
        expect(result.id).toBe(surveySchema.id)
        expect(result.userId).toBe("hashValue") 
        expect(result).toMatchSnapshot()
    });
    
    it("Will output JSON", async () => {
        engineOptions.outputFormat = "json";
        
        const engine = new EnquirerEngine(engineOptions);
        
        const result: string = await engine.run(surveySchema) as string
        
        const parsedResult = JSON.parse(result);
        expect(parsedResult.id).toBe(surveySchema.id)
        expect(parsedResult.userId).toBe(engineOptions.userId)
        
        expect(result).toMatchSnapshot()
    });
    
    it("will default to raw json if no output is provided", async () => {
        engineOptions.outputFormat = undefined;
        const engine = new EnquirerEngine(engineOptions);

        const result: IEngineOutput = await engine.run(surveySchema) as IEngineOutput

        expect(result.id).toBe(surveySchema.id)
        expect(result.userId).toBe(engineOptions.userId)
        expect(result).toMatchSnapshot()

    });
    it("Will output CSV", async () => {
        engineOptions.outputFormat = "csv";
        
        const engine = new EnquirerEngine(engineOptions);
        
        const result: string = await engine.run(surveySchema) as string

        expect(result).toMatchSnapshot()
    });
})