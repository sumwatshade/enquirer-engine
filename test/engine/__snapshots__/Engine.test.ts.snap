// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Engine Will default a user ID if none are provided 1`] = `
Object {
  "id": "my-test-schema",
  "respondedAt": "2015-06-14T22:13:45.275Z",
  "results": Array [
    Object {
      "choice": 2,
      "id": "tacobell",
    },
    Object {
      "choice": 1,
      "id": "pizza",
    },
    Object {
      "choice": 3,
      "id": "inout",
    },
    Object {
      "choice": 2,
      "id": "tacobell2",
    },
    Object {
      "choice": 1,
      "id": "pizza2",
    },
    Object {
      "choice": 3,
      "id": "inout2",
    },
  ],
  "userId": "hashValue",
}
`;

exports[`Engine Will output CSV 1`] = `
"# Survey ID: my-test-schema
# Responded at: 2015-06-14T22:13:45.275Z
# User: test-user
promptId, result
tacobell, 2
pizza, 1
inout, 3
tacobell2, 2
pizza2, 1
inout2, 3"
`;

exports[`Engine Will output JSON 1`] = `"{\\"id\\":\\"my-test-schema\\",\\"userId\\":\\"test-user\\",\\"respondedAt\\":\\"2015-06-14T22:13:45.275Z\\",\\"results\\":[{\\"id\\":\\"tacobell\\",\\"choice\\":2},{\\"id\\":\\"pizza\\",\\"choice\\":1},{\\"id\\":\\"inout\\",\\"choice\\":3},{\\"id\\":\\"tacobell2\\",\\"choice\\":2},{\\"id\\":\\"pizza2\\",\\"choice\\":1},{\\"id\\":\\"inout2\\",\\"choice\\":3}]}"`;

exports[`Engine Will run a scale-based survey 1`] = `
Object {
  "id": "my-test-schema",
  "respondedAt": "2015-06-14T22:13:45.275Z",
  "results": Array [
    Object {
      "choice": 2,
      "id": "tacobell",
    },
    Object {
      "choice": 1,
      "id": "pizza",
    },
    Object {
      "choice": 3,
      "id": "inout",
    },
    Object {
      "choice": 2,
      "id": "tacobell2",
    },
    Object {
      "choice": 1,
      "id": "pizza2",
    },
    Object {
      "choice": 3,
      "id": "inout2",
    },
  ],
  "userId": "test-user",
}
`;

exports[`Engine will default to raw json if no output is provided 1`] = `
Object {
  "id": "my-test-schema",
  "respondedAt": "2015-06-14T22:13:45.275Z",
  "results": Array [
    Object {
      "choice": 2,
      "id": "tacobell",
    },
    Object {
      "choice": 1,
      "id": "pizza",
    },
    Object {
      "choice": 3,
      "id": "inout",
    },
    Object {
      "choice": 2,
      "id": "tacobell2",
    },
    Object {
      "choice": 1,
      "id": "pizza2",
    },
    Object {
      "choice": 3,
      "id": "inout2",
    },
  ],
  "userId": "test-user",
}
`;
